# duplexseq
Scripts for analysis of duplex sequencing data


## Overview

The scripts in this repository are used for processing of raw reads generated by [the duplex sequencing method](https://www.nature.com/articles/nprot.2014.170) and subsequent variant calling and characterization.

## Dependencies

- [BBMerge](https://jgi.doe.gov/data-and-tools/bbtools/bb-tools-user-guide/bbmerge-guide/)
- [cutadapt](https://cutadapt.readthedocs.io/en/stable/)
- [bowtie2](http://bowtie-bio.sourceforge.net/bowtie2/index.shtml)
- [sloan.pm Perl module](https://github.com/dbsloan/perl_modules)
- Perl
- Python3 (if using a k-mer database to screen variants with `duplexseq_sam_to_indels.pl` or `duplexseq_sam_to_snps.pl`)

## Pipeline Summary

**Raw read processing and filtering of read family consensus sequence**
- `duplexseq_process_reads.pl`
- `duplexseq_family_filter_and_trim.pl`

**Mapping to reference genome**
- `bowtie2`

**Raw variant calling**
- `duplexseq_sam_to_indels.pl`
- `duplexseq_sam_to_snps.pl`

**Aggregating data from multiple libraries, variant filtering, and summary**
- `duplexseq_indels_post_processing.pl`
- `duplexseq_snps_post_processing.pl`
- `duplexseq_coverage_summary.pl`

**Other scripts**
- `duplexseq_batchscripts.pl` (generates a shell script for each input library to automate the first 3 steps above under our standard parameters)
- `duplexseq_indels_kmercount.py` and `duplexseq_snps_kmercount.py` (accessory scripts called during SNP and indel post-processing if using a k-mer database)
- `duplexseq_snp_classification.pl` (functional annotation of filtered SNP calls)